contributor-docs/adrs/adr-001-custom-elements.md

```diff
@@ -0,0 +1,115 @@
+# Custom element usage in Primer React Components ADR
+
+## Status
+
+## Related Documents
+- April 18th, 2018: [Custom Elements Example](https://github.com/primer/components/pull/13)
+- May 8th, 2018:  [Custom Elements Experiment](https://github.com/primer/components/pull/14)
+- May 18th 2018: [React & Custom Elements](https://docs.google.com/document/d/1b6D2jW3ztQJiKEtQINbb4JoHJ3F3szU6lBX-fGce6aY/edit?usp=sharing)
+- October 16th 2019: [Issue exploring pros & cons of reusing .com JS in React](https://github.com/github/issues-index-experiment/issues/90)
+- November 15th 2019: [Experiment porting PRC to web components](https://github.com/github/ui-engineering/issues/12#issuecomment-558257842)
+- November 27th 2019: [Usage of custom elements in IIE project](https://github.com/github/issues-index-experiment/pull/133)
+- November 27th 2019: [Styling custom elements with styled components](https://github.slack.com/archives/CMZ4DC9BL/p1574883198055600)
+- June 9th 2020: [Discussion in Slack about using skatejs for SSR](https://github.slack.com/archives/C0ER2LCG2/p1591707104461300)
+- November 10th 2020: [Replace dialog with details-dialog](https://github.com/primer/components/issues/907)
+
+## Context
+Throughout the last few years folks from the Design Infrastructure, Web Systems & UI Platform teams have discussed the idea of using custom elements for behaviors in Primer React Components. The main goal of using custom elements in PRC is to be able to author behaviors once and reuse them in any framework. Several experiments have been conducted which are listed above.
+
+## Assumptions
+De-duplication is not our highest or only priority. Attempts at de-duplication must be weighed against changes to the maintainer, developer, and customer experience.
+
+## Findings
+
+### Developer Experience Regressions
+- Rendering custom elements in SSR applications is a bit tricky and requires a third party framework such as [@skatejs/ssr](https://github.com/skatejs/skatejs/tree/master/packages/ssr). This is a hit to developer experience because it would require consumers of PRC to set up the service, PRC won’t work right out of the box (in applications using SSR, such as all of our Doctocat sites) without it if we use custom elements in the library.
```

Comment:
Nit: A third-party framework isn’t necessarily required; I wrote a ~50 line [function](https://github.com/github/ui-eng-framework-survey/blob/6a8ec7c2080146c7495c025500c96cc7287cbb20/utils.ts#L50-L111) for rendering custom elements server-side for `github/ui-eng-framework-survey`

Reply:
Custom Elements are a fundamentally different design pattern to React, which I think is what is causing confusion here. Additionally the term "SSR" means many different things, but they're conflated into one with React.

To borrow a term used in the React ecosystem, React components are designed with _isomorphism_ as an ideal while Custom Elements are the opposite. The `class extends HTMLElement {}` part of a custom element is designed to _only run in the browser execution environment_. This means They shouldn't be imported and run on the server. You can try to do that by loading a DOM library like `jsdom` into the server, but it's attempting to solve a problem which shouldn't be solved.

This doesn't mean Custom Elements cannot be rendered server side, of course. The HTML part of Custom Elements is just that, HTML. It'll work anywhere that can output a string. Bash can be used as a server side language and output Custom Elements, just like React, just like Go, just like C, just like Rails. HTML is the primitive which is _platform agnostic_ and the server emits HTML tags that match the names of the custom elements, and the Browser will combine the JS logic to the HTML tag. This is progressive enhancement. It means you can ship custom elements that don't require JS to function.

A big part of the "SSR story" is rendering subtrees. Again, because Custom Elements are platform agnostic they leave rendering of subtrees to the user. For example our `<details-dialog>` component expects to be rendered as a sub-element of `<details>`, so the _implicit_ expectation is that in whatever server side language you're using, you write a component that does that wiring. In the case of the monolith we use ViewComponents for that. React could be used too. This grants flexibility in the Server Side implementation without tying you to JS. Does it mean writing two implementations for a component? No (unless you're trying hard to do that). Custom Elements strongly encourage progressive enhancement, and technology agnostic to servers. The JS _shouldn't_ be doing hydration, subtreets, etc, if you can avoid it.

If you _want_ to render SubTrees in Custom Elements, it is possible using Shadow DOM. Server side rendering of ShadowDOM is still nascent, but the [Declarative Shadow DOM spec](https://web.dev/declarative-shadow-dom/) will solve this. It's not shipping in browsers yet, but it can be polyfilled with a few lines of JS. [Catalyst already ships a polyfill for declarative shadowDOM templates that is 11LOC](https://github.com/github/catalyst/blob/main/src/auto-shadow-root.ts). Declarative Shadow DOM is again HTML only, so is not coupled to JS or any framework, React can output Declarative ShadowDOM contents similar to how React might write a component itself:

```js
const ReactProgressBar = ({progress, bg, theme, ...rest}) => {
  return (
    <ProgressContainer theme={theme} {...rest}>
      <Bar progress={progress} bg={bg} theme={theme} />
    </ProgressContainer>
  )
}

const ShadowDOMProgressPar = ({progress, bg, theme, ...rest}) => {
 return (
    <progress-container theme={theme} {...rest}>
      <template shadowroot="open">
        <bar-element progress={progress} bg={bg} theme={theme} />
      </template>
    </progress-container>
  )
}